replica/database.cc

```diff
@@ -1808,13 +1808,22 @@ future<> database::apply_in_memory(const frozen_mutation& m, schema_ptr m_schema
 
     data_listeners().on_write(m_schema, m);
 
```

Comment:
> median 1580850.76 tps ( 70.3 allocs/op,  14.2 tasks/op,   54240 insns/op,        0 errors)

12 extra allocations for 500 insns/ops? it doesn't add up.

Reply:
I'm not sure exactly where they come from.
I suspect it's due to unfreezing outside the context of the memtable allocating_section.
Is it possible that those allocation weren't accounted for here:
```
memtable::apply(const frozen_mutation& m, const schema_ptr& m_schema, db::rp_handle&& h) {
    with_allocator(allocator(), [this, &m, &m_schema] {
        _table_shared_data.allocating_section(*this, [&, this] {
            auto& p = find_or_create_partition_slow(m.key());
            mutation_partition mp(*m_schema);
            partition_builder pb(*m_schema, mp);
            m.partition().accept(*m_schema, pb);
```

This code effectively unfreezes the frozen_mutation into `mp`, the difference is that it's done inside the allocating_section.

Reply:
The report doesn't make sense. If the number of allocations increases by 12, I expect a large increase in the number of instructions.

Reply:
I suspect it might be an accounting problem that is just exposed here, but I can't prove this theory yet.

Reply:
@avikivity the extra allocations count comes from the fact the unfreezing is done outside of the memtable `allocating_section` - since we count only seastar mallocs and not lsa allocs.

With the following patch perf-simple-query --write reports 8 more allocations:
```diff
diff --git a/replica/memtable.cc b/replica/memtable.cc
index 77d1a95f9a..25199b37d1 100644
--- a/replica/memtable.cc
+++ b/replica/memtable.cc
@@ -796,13 +796,13 @@ memtable::apply(const mutation& m, db::rp_handle&& h) {
 
 void
 memtable::apply(const frozen_mutation& m, const schema_ptr& m_schema, db::rp_handle&& h) {
-    with_allocator(allocator(), [this, &m, &m_schema] {
+    mutation_partition mp(*m_schema);
+    partition_builder pb(*m_schema, mp);
+    m.partition().accept(*m_schema, pb);
+    _stats_collector.update(*m_schema, mp);
+    with_allocator(allocator(), [&] {
         _table_shared_data.allocating_section(*this, [&, this] {
             auto& p = find_or_create_partition_slow(m.key());
-            mutation_partition mp(*m_schema);
-            partition_builder pb(*m_schema, mp);
-            m.partition().accept(*m_schema, pb);
-            _stats_collector.update(*m_schema, mp);
             p.apply(region(), cleaner(), *_schema, std::move(mp), *m_schema, _table_stats.memtable_app_stats);
         });
     });
```

Reply:
It makes sense that there are 8 extra allocations. But then I'd expect around 1600 extra instructions. Maybe my expectations are incorrect, but something smells here.

Reply:
1600 extra instructions from trading lsa allocations with seastar allocations?

I'll try to count lsa allocations in perf-simple-query in addition to the seastar allocations so we can see the whole picture.